lib/saluki-components/src/destinations/datadog_metrics/endpoint/endpoints.rs

```diff
@@ -0,0 +1,149 @@
+use snafu::{ResultExt, Snafu};
+use url::Url;
+
+use crate::destinations::datadog_metrics::DEFAULT_SITE;
+
+/// A set of additional API endpoints to forward metrics to.
+///
+/// Each endpoint can be associated with multiple API keys. Requests will be forwarded to each unique endpoint/API key pair.
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct AdditionalEndpoints(HashMap<String, Vec<String>>);
+
+/// Error type for invalid endpoints.
+#[derive(Debug, Snafu)]
+#[snafu(context(suffix(false)))]
+pub enum EndpointError {
+    Parse { source: url::ParseError, domain: String },
+}
+/// A single API endpoint and its associated API keys.
+#[derive(Default)]
+pub struct SingleDomainResolver {
+    domain: String,
+    api_keys: Vec<String>,
+}
+
+impl SingleDomainResolver {
+    /// Returns the domain of the resolver.
+    pub fn domain(&self) -> &String {
+        &self.domain
+    }
+
+    /// Returns the API keys associated with the domain.
+    pub fn api_keys(&self) -> &Vec<String> {
+        &self.api_keys
+    }
+
+    fn add_api_key(&mut self, api_key: String) {
+        self.api_keys.push(api_key);
+    }
+
+    fn set_domain(&mut self, domain: String) {
+        self.domain = domain;
+    }
+}
+
+/// Converts [`AdditionalEndpoints`] into a list of [`SingleDomainResolver`]s.
+pub fn create_single_domain_resolvers(
+    endpoints: &AdditionalEndpoints,
+) -> Result<Vec<SingleDomainResolver>, EndpointError> {
+    let mut resolvers = Vec::new();
+    for (domain, api_keys) in &endpoints.0 {
+        let new_domain = add_adp_version_to_domain(domain.to_string())?;
+        let mut seen = HashSet::new();
+        let mut resolver = SingleDomainResolver::default();
+        for api_key in api_keys {
+            let trimmed_api_key = api_key.trim();
+            if trimmed_api_key.is_empty() || seen.contains(trimmed_api_key) {
+                continue;
+            }
+            seen.insert(trimmed_api_key);
+            resolver.add_api_key(trimmed_api_key.to_string());
+        }
+        if !resolver.api_keys.is_empty() {
+            resolver.set_domain(new_domain.trim_end_matches("/").to_string());
+            resolvers.push(resolver);
+        }
+    }
+    Ok(resolvers)
+}
+
+impl FromStr for AdditionalEndpoints {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let endpoints: HashMap<String, Vec<String>> = serde_json::from_str(s)?;
+        Ok(AdditionalEndpoints(endpoints))
+    }
+}
+
+/// Return the correct prefix for a domain.
+pub fn get_domain_prefix(app: String) -> String {
+    let app_details = saluki_metadata::get_app_details();
+    let version = app_details.version();
+    format!(
+        "{}-{}-{}-{}.agent",
+        version.major(),
+        version.minor(),
+        version.patch(),
+        app
+    )
+}
+
+/// Prefixes the domain with the ADP version.
+pub fn add_adp_version_to_domain(domain: String) -> Result<String, EndpointError> {
```

Comment:
I think we _probably_ just want some simple unit tests for this?

At some point, it's very likely we'll want to refactor this code to use it generically for handling all types of different DD intake endpoints, and right now, it's very specific to metrics (i.e. all endpoints need to be `app.<site>`) so having some simple unit tests will obviously help us avoid breaking the code in the future, but make it a little more clear what it is does and what it expects the input to look like.